### ISSUE TITLE
Accessing a private policy repo using the conftest Docker image is not working

### ISSUE THREAD

**Issue by tspearconquest** [open-policy-agent/conftest#567]
Hi,

I have a use case where we are running conftest in our GitLab CI and have some problems retrieving the policies from our private repo due to using conftest within a DinD container.

In this CI we have a pre-test stage configured to pull down policies from our private repo and run conftest with them against our Dockerfile for building a privately hosted copy of conftest for us to use in order to avoid dockerhub ratelimits. This takes place inside of a gitlab-runner docker container. This pre-test stage works fine. We login with docker in the gitlab-runner container, and conftest uses the token to connect to our private repo.

In a later stage after the build is completed, we push up the private image and then run it within a DinD container.
This fails however, because the contfest binary has no option to pass login details. I did try to pass my config file over a volume mount into the container but I don't see a way to bring in the token itself.

So, what we need is a way to either pass the token from the parent container to the DinD child container, or a way to tell conftest to login, something similar to: `conftest login --user someuser --password-stdin example.com/some_registry <somefile` or another similar command which can login to our registry in order to pull the policies while bypassing the docker token.

**Comment by jpreese** [comment]
Which container registry are you trying to pull from? Under the hood Conftest just uses [go-getter](https://github.com/hashicorp/go-getter) to push and pull the policies so any mode of auth (token, environment variable, etc) should be available to you.

For example, with GCP you could set the `GOOGLE_APPLICATION_CREDENTIALS` environment variable.

**Comment by tspearconquest** [comment]
It's GitLab. I think I should reiterate that this only happens in a specific scenario.

1. Pipeline starts a container
2. `docker login` takes place inside container
3. `docker run open-policy-agent/contest pull $policies`
4. Child container starts and contest gets a 403 because the token is not passed to the child container.

So my question is how do we pass a token into the container run during step 3?

**Comment by jpreese** [comment]
This doesn't feel specific to Conftest and is more about adding environment variables and mounting configurations into containers that are in DinD mode. I don't have a lot of DinD experience, but if you've got a configuration file or PAT you should be able to set the environment variable and/or mount what you need for Conftest to read it. Something like (pseudo-ish):

`docker run -v ~/token:./tmp -e DOCKER_CONFIG_PATH=/tmp open-policy-agent/conftest pull`

Alternatively since you're just running the conftest Dockerfile and not a suite of tools, could you just download Conftest in your pipeline? Typically pipelines have a tools cache that won't download the binary every time.

**Comment by tspearconquest** [comment]
Hi John,

To your last point regarding just downloading the binary, we do that in the first stage.
The stage where we are having issues is when we are trying to validate that the image works.

We plan to use this image in other pipelines, and so we would be calling this image in all of our pipelines across many projects in the organization in this same way. We are doing this with a few other open source containerized projects without any issues, however those projects don’t require direct access to the repo to download policies.

The solution I made last night is to download the policies using the conftest binary, and then pass those in via a volume mount:

Pseudo:
```
# wget conftest
# chmod +x conftest
# ./conftest pull $policies
# docker login foo
# docker run -v /path/to/policies:/policy our_repo/conftest:latest test Dockerfile -p /policy --all-namespaces
```

This seems to be working, but it’s not the cleanest looking CI pipeline code, my colleagues are confused about it so I was hoping to make it cleaner.

Thanks

From: John Reese ***@***.***>
Date: Tuesday, May 18, 2021 at 8:05 AM
To: open-policy-agent/conftest ***@***.***>
Cc: Thomas Spear ***@***.***>, Author ***@***.***>
Subject: Re: [open-policy-agent/conftest] Conftest pull from private repo in DinD (#567)

CAUTION: This email originated from outside the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.

This doesn't feel specific to Conftest and is more about adding environment variables and mounting configurations into containers that are in DinD mode. I don't have a lot of DinD experience, but if you've got a configuration file or PAT you should be able to set the environment variable and/or mount what you need for Conftest to read it. Something like (pseudo-ish):

docker run -v ~/token:./tmp -e DOCKER_CONFIG_PATH=/tmp open-policy-agent/conftest pull

Alternatively since you're just running the conftest Dockerfile and not a suite of tools, could you just download Conftest in your pipeline? Typically pipelines have a tools cache that won't download the binary every time.

**Comment by tspearconquest** [comment]
Also, for security reasons we need to keep a local copy in our own repo of any open source software we use.

Typically, we add some layers on top of an upstream docker image in order to patch the base operating system image underlying the container, or we pull some binary out of a tarball into an Alpine container which is what we've done with conftest.

We are trying to leverage the container in our Pipeline now and that's the issue we are having because conftest doesn't appear to support connecting to a private repo from within a docker container. This would happen regardless of whether I was running it with DinD, or even if I ran this container on my local machine...

The requirement to download the conftest binary independently from the docker container I created to harness the conftest binary, is an issue we need to see resolved. We should be able to run the container twice as follows, but we cannot:

```
# docker run --rm -v /path/to/policy:/policy ${CONFTEST_IMAGE} pull ${POLICIES_FROM_PRIVATE_REPO}
-- here we get an error --
# docker run --rm -v /path/to/policy:/policy ${CONFTEST_IMAGE} test Dockerfile -p /policy --all-namespaces

```

Whereas if we do as I demonstrated in my previous comment, it works fine because it uses the token already setup by the docker login command.

**Comment by tspearconquest** [comment]
We tried your steps without success.

We are able to use the container in 2 cases...

Case 1 is the one I described two comments earlier.
Case 2 I will describe below:

We modified the Dockerfile to pull in the docker binary from apk.
Then we modified the CI job to use the conftest image directly instead of a docker image (bypassing docker in docker), and so within that image we are calling docker login, followed by conftest pull, followed by conftest test.

This works fine, it's less ideal because it's different from how the other pipelines are structured, but it works.
Ideally, we would like the ability to pass authentication details as user and password environment variables for the image and conftest would read those variables directly in order to pass the details over to the server its connecting to.

One thought I had was to try doing the conftest pull via http and passing http style username and password values. No idea if it will work...

**Comment by jpreese** [comment]
Glad you got at least _something_ working! I would have to do some testing on my end to fully understand the problem if you're still running into issues with DinD.. 

Conftest isn't doing anything special with the credentials and relies on `go-getter` to find the auth. If the DinD container has everything set up properly (token mounted, environment variables set, etc) it shouldn't be any different as far as I'm aware.

**Comment by tspearconquest** [comment]
I can clarify further and as I mentioned we can reproduce this issue without DinD at all.

The failure can be reproduced on my laptop with the below commands
```
host_shell$ docker login registry.gitlab.com
host_shell$ docker run --rm -it open-policy-agent/conftest:latest pull registry.gitlab.com/private_repo/conftest_policies
```
Here we launch the conftest container (note this is the upstream conftest container) and simply try to pull the conftest policies from our own access-restricted repo with the default entrypoint of the image being the /conftest command.

There is no method I could find to pass the authentication token from that first docker login command on the host system into the container where conftest is running, and this is the issue I am trying to get resolved.

My workaround works because instead of doing the above, we are instead doing (effectively) this:
```
host_shell$ docker login registry.gitlab.com
host_shell$ docker run --rm -it --entrypoint /bin/sh registry.gitlab.com/private_repo/conftest:latest
# docker login -u registry_user -p registry_password
# /conftest pull registry.gitlab.com/private_repo/conftest_policies
```
It's a bit more nuanced than that, because of how CI pipelines work, but the above examples should hopefully help to demonstrate the issue more effectively without having to dive into DinD at all.

Our private repo copy of conftest used above has a new layer we added to the base image you provide upstream so that we could run the `docker login` command inside of the conftest container, to allow the above workaround to operate properly.

**Comment by jpreese** [comment]
From your first code snippet, you're only mounting the docker sock -- not the credentials that were generated when calling `docker login`.

You'd have to do something like:
```
docker run --rm -it -v /var/run/docker.sock:/var/run/docker.sock -v ${HOME}/.docker/config.json:/root/.docker open-policy-agent/conftest:latest pull registry.gitlab.com/private_repo/conftest_pol
```

**Comment by tspearconquest** [comment]
We tried that as well. No luck.

![image](https://user-images.githubusercontent.com/81998567/120042004-ec168880-bfce-11eb-9432-37163e73b361.png)

```
CONFTEST_IMAGE: registry.gitlab.com/conquestcyber/devops/conftest/conftest:0.25.0
REGO_POLICIES: registry.gitlab.com/conquestcyber/devops/rego-policies
```

**Comment by jpreese** [comment]
Is your private Conftest image the same image thats pushed to DockerHub at the moment? Or did you add any custom steps

**Comment by tspearconquest** [comment]
It is actually a custom image.

```docker
FROM alpine:3.13.5 as builder

ARG CONFTEST_VER

RUN apk update && \
    apk add --no-cache wget git docker && \
    wget https://github.com/open-policy-agent/conftest/releases/download/v${CONFTEST_VER}/conftest_${CONFTEST_VER}_Linux_x86_64.tar.gz && \
    tar xzf conftest_${CONFTEST_VER}_Linux_x86_64.tar.gz && \
    mv conftest /usr/bin && \
    rm -f conftest_${CONFTEST_VER}_Linux_x86_64.tar.gz LICENSE README.md && \
    adduser -D -u 65532 conftest && \
    mkdir /project
USER 65532
WORKDIR /project
ENTRYPOINT [ "conftest" ]
```

**Comment by jpreese** [comment]
You're setting `USER` to 65532, but mounting the docker creds to `roots` home directory. Docker looks for the credentials in $HOME/.docker/config.json unless you override it with $DOCKER_CONFIG.

So you'd either need to change where the `config.json` is mounted to inside of the container or pass in `-e DOCKER_CONFIG` with a different directory.

**Comment by tspearconquest** [comment]
Thanks, I tried a slightly different setup that should still work but still got the same error.

```docker
FROM alpine:3.13.5 as builder

ARG CONFTEST_VER
ENV DOCKER_CONFIG=/project/.config

RUN apk update && \
    apk add --no-cache wget git docker && \
    wget https://github.com/open-policy-agent/conftest/releases/download/v${CONFTEST_VER}/conftest_${CONFTEST_VER}_Linux_x86_64.tar.gz && \
    tar xzf conftest_${CONFTEST_VER}_Linux_x86_64.tar.gz && \
    mv conftest /usr/bin && \
    rm -f conftest_${CONFTEST_VER}_Linux_x86_64.tar.gz LICENSE README.md && \
    adduser -D -u 65532 conftest && \
    mkdir /project
USER 65532
WORKDIR /project
ENTRYPOINT [ "conftest" ]
```

Followed by `$ docker run --rm -v "$(pwd)":/project -v /root/.docker/config.json:/project/.docker/config.json -v /var/run/docker.sock:/var/run/docker.sock $CONFTEST_IMAGE pull $REGO_POLICIES`

**Comment by jpreese** [comment]
The `ENV` is set to `/project/.config` it should be `/project/.docker` based on your `docker run` statement.

**Comment by tspearconquest** [comment]
Apologies, you are right. I copied from the pre-edited Dockerfile and added that manually, but then when I entered it in the Dockerfile itself, it was properly entered. The failure happened with the environment set correctly.

That said, I've edited the above comment now.

**Comment by jpreese** [comment]
And as a sanity check you've verified that the `config.json` exists in the container under /project/.docker? The `-v` flag will create the file regardless of whether or not it exists on the host.

**Comment by tspearconquest** [comment]
Any attempt to mount the file or the directory has been futile thus far.

```
docker run -v /root/.docker/config.json:/project/.docker/config.json
```
With this one, I could see /project/.docker/config.json was an empty directory inside the container, rather than the file from the host. I did verify the file is present on the host.

```
docker run -v /root/.docker/config.json:/project/.docker
docker run -v /root/.docker:/project/.docker
```
Both of these resulted in /project/.docker being an empty directory inside the container

It appears that Docker does not want us to do that. So would it be possible to add in flags to conftest that can pass the username and password to go-getter?

**Comment by tspearconquest** [comment]
In case its permissions related, I am running some further tests.

**Comment by tspearconquest** [comment]
Even copying the config.json file to /tmp, making it world readable, and then trying to mount it from there is giving the same results as above.

```
$ ls -laR /root/.docker
/root/.docker:
total 12
drwxr-xr-x    2 root     root          4096 May 30 00:38 .
drwx------    1 root     root          4096 May 30 00:38 ..
-rw-r--r--    1 root     root           174 May 30 00:38 config.json
$ cp /root/.docker/config.json /tmp
$ chmod 644 /tmp/config.json
$ docker run --rm -v "$(pwd)":/project -v /tmp/config.json:/project/.docker/config.json -v /var/run/docker.sock:/var/run/docker.sock --entrypoint "/bin/sh" "${CONFTEST_LATEST}" -c "ls -laR /project/.docker"
-snip-
Status: Downloaded newer image for registry.gitlab.com/our_repo/conftest/conftest:latest
/project/.docker:
total 8
drwxr-xr-x    3 root     root          4096 May 30 00:54 .
drwxrwxrwx    5 root     root          4096 May 30 00:54 ..
drwxr-xr-x    2 root     root            40 May 30 00:54 config.json
/project/.docker/config.json:
total 4
drwxr-xr-x    2 root     root            40 May 30 00:54 .
drwxr-xr-x    3 root     root          4096 May 30 00:54 ..
```

**Comment by jpreese** [comment]
You could try moving the `config.json` from `/project/.docker` to another location on disk because you're first mounting `pwd` into `/project`. This _should_ be OK, but depending on how your CI's filesystem is laid out could be problematic -- worthwhile to remove that potential point of failure though. There's lots of documentation and discussions out there talking about potential reasons why your mounted files are empty.

**Comment by tspearconquest** [comment]
I did as you suggested, and moved it to /etc/conquest, which should work fine. Results were the same.
I must contest that this is not a mounted files are "empty" issue but rather the mounted file is being presented inside the container as a directory instead of a file. Quite a different scenario imho, and still not the root cause of the problem.
Thus far, I've presented quite a bit of evidence that this is not working, but have not seen any evidence from you demonstrating that this works in your own testing.

Do you mind sharing what you're doing on your end where it **is** working, so that I can try to reproduce *that* setup on my end and work backwards from there?

**Comment by jpreese** [comment]
I'm not running a DinD setup with GitLab. When I run a DinD instance on my own machine and mount the files in, I see them. You could also try asking the GitLab community to see how they're running similar setups where DinD setups require auth.

**Comment by tspearconquest** [comment]
Thanks but seeing the files and being able to retrieve the policies are not the same thing.
What I want to try to do is reproduce your setup locally on my laptop and try to get even *that* working (its not currently) and work the problem from there. Its clear I'm doing something wrong, I don't deny that, and so I'm just trying to ascertain what it is that needs adjustment between your local environment and mine.

**Comment by tspearconquest** [comment]
Just to clarify, I am asking if you have tested pulling from a private repository using your docker image. If you haven't tested due to not having a private repo, then this is an unsupported setup, and you should say so. If you are interested in testing this further and *need* a repo to test with, I don't mind helping out with that.

Ultimately, I would like to get to the point that the below command (or a variant of it with more flags) works properly on my laptop as you imply that it should work. That would be a great place for me to continue troubleshooting this from, because currently it is not working:

`docker run openpolicyagent/conftest:latest pull $private_policy_repo_path`

**Comment by tspearconquest** [comment]
With some further testing, I was able to get the config.json into the container.
Unfortunately it didn't work.

Strangely, it works when I used the image from dockerhub.

Looking through our Dockerfile and comparing it to yours, I realized we were generating more or less the same image as what your Dockerfile does, with the exception that ours doesn't build conftest from source, and doesn't work with our private repo while yours does both, so I have redesigned our Dockerfile to add a layer on top of your image instead and it seems to be working well now.

We can consider the issue resolved.
